src/evaluators/ForStatement.js

```diff
@@ -268,13 +280,151 @@ function ForBodyEvaluation(
   }
 }
 
+let BailOutWrapperClosureRefVisitor = {
+  ReferencedIdentifier(path: BabelTraversePath, state: BailOutWrapperInfo) {
+    if (path.node.name === "arguments") {
+      state.usesArguments = true;
+    }
+  },
+  ThisExpression(path: BabelTraversePath, state: BailOutWrapperInfo) {
+    state.usesThis = true;
+  },
+  "BreakStatement|ContinueStatement"(path: BabelTraversePath, state: BailOutWrapperInfo) {
+    if (path.node.label !== undefined) {
+      state.usesGotoToLabel = true;
+    }
+  },
+  ReturnStatement(path: BabelTraversePath, state: BailOutWrapperInfo) {
+    state.usesReturn = true;
+  },
+};
+
+function generateRuntimeForStatement(
+  ast: BabelNodeForStatement,
+  strictCode: boolean,
+  env: LexicalEnvironment,
+  realm: Realm,
+  labelSet: ?Array<string>
+): AbstractValue {
+  let wrapperFunction = new ECMAScriptSourceFunctionValue(realm);
+  let body = t.blockStatement([ast]);
+  ((body: any): FunctionBodyAstNode).uniqueOrderedTag = realm.functionBodyUniqueTagSeed++;
+  wrapperFunction.$ECMAScriptCode = body;
+  wrapperFunction.$FormalParameters = [];
+  wrapperFunction.$Environment = env;
+  // We need to scan to AST looking for "this", "return", labels and "arguments"
```

Comment:
Why not throw?

Reply:
Throw where and with what? Sorry, not sure I get what you mean here.

Reply:
If the loop body throws, it terminates early. That seems analogues to a loop body with a return in it. I don't see why you check for the latter, but not the former.

Reply:
I understand now. Thanks, added that case.